File: js/textNormalizer.js
/**
 * A class for normalizing text containing unusual writing styles,
 * such as repeated characters, text in asterisks, or
 * unusual capitalization.
 */
class TextNormalizer {
    /**
     * Normalizes text by applying multiple normalization rules.
     *
     * @param {string} input - The text to normalize.
     * @returns {string} The normalized text.
     */
    normalizeText(input) {
        if (!input) {
            return input;
        }

        let result = input;
        
        // Normalize stretched words (e.g., "hellooo" -> "hello")
        result = this.normalizeRepeatedCharacters(result);
        
        // Remove text in asterisks (e.g., "*cough*")
        result = this.removeAsterisks(result);
        
        // Normalize capitalization (e.g., "HeLLo" -> "Hello")
        result = this.normalizeCasing(result);
        
        return result;
    }

    /**
     * Normalizes repeated characters in text.
     *
     * @param {string} input - The text to normalize.
     * @returns {string} The text with normalized repeated characters.
     */
    normalizeRepeatedCharacters(input) {
        if (!input) {
            return input;
        }

        // Define word boundaries to identify words
        const pattern = /\b\w+\b/g;
        
        return input.replace(pattern, (word) => {
            let normalized = '';
            
            for (let i = 0; i < word.length; i++) {
                const current = word[i];
                normalized += current;
                
                // Skip all subsequent identical characters
                while (i + 1 < word.length && word[i + 1] === current) {
                    i++;
                }
            }
            
            return normalized;
        });
    }

    /**
     * Removes text surrounded by asterisks and fixes any resulting extra spaces.
     *
     * @param {string} input - The text to normalize.
     * @returns {string} The text without asterisk content and no extra spaces.
     */
    removeAsterisks(input) {
        if (!input) {
            return input;
        }
        
        // Remove text between asterisks completely
        let result = input.replace(/\*[^*]*\*/g, '');
        
        // Fix any resulting double spaces or more
        result = result.replace(/\s{2,}/g, ' ');
        
        return result;
    }

    /**
     * Normalizes capitalization of words.
     *
     * @param {string} input - The text to normalize.
     * @returns {string} The text with normalized capitalization.
     */
    normalizeCasing(input) {
        if (!input) {
            return input;
        }

        // Define word boundaries to identify words
        const pattern = /\b\w+\b/g;
        
        return input.replace(pattern, (word) => {
            // If the word has mixed capitalization
            if (this.hasMixedCasing(word)) {
                // Convert everything to lowercase
                let normalized = word.toLowerCase();
                
                // First letter uppercase if the original was uppercase
                if (/^[A-Z]/.test(word)) {
                    normalized = normalized.charAt(0).toUpperCase() + normalized.slice(1);
                }
                
                return normalized;
            }
            
            return word;
        });
    }

    /**
     * Checks if a word contains mixed capitalization.
     *
     * @param {string} word - The word to check.
     * @returns {boolean} True if the word has mixed capitalization, otherwise False.
     */
    hasMixedCasing(word) {
        if (word.length <= 1) {
            return false;
        }
        
        let hasUpper = false;
        let hasLower = false;
        
        for (let i = 0; i < word.length; i++) {
            const char = word[i];
            
            if (char >= 'A' && char <= 'Z') {
                hasUpper = true;
            } else if (char >= 'a' && char <= 'z') {
                hasLower = true;
            }
            
            if (hasUpper && hasLower) {
                return true;
            }
        }
        
        return false;
    }
}

// Example usage
// const normalizer = new TextNormalizer();
// const normalizedText = normalizer.normalizeText("Hellooo how areee you *cough* today?");
// console.log(normalizedText);  // "Hello how are you today?"

// For Node.js environment
if (typeof module !== 'undefined' && module.exports) {
    module.exports = TextNormalizer;
}

// For browser environment and ES modules
if (typeof window !== 'undefined') {
    window.TextNormalizer = TextNormalizer;
}

File: csharp/TextNormalizer.cs
using System;
using System.Text;
using System.Text.RegularExpressions;

namespace TextProcessing
{
    /// <summary>
    /// A class for normalizing text containing unusual writing styles,
    /// such as repeated characters, text in asterisks, or
    /// unusual capitalization.
    /// </summary>
    public class TextNormalizer
    {
        /// <summary>
        /// Normalizes text by applying multiple normalization rules.
        /// </summary>
        /// <param name="input">The text to normalize.</param>
        /// <returns>The normalized text.</returns>
        public string NormalizeText(string input)
        {
            if (string.IsNullOrEmpty(input))
                return input;

            string result = input;
            
            // Normalize stretched words (e.g., "hellooo" -> "hello")
            result = NormalizeRepeatedCharacters(result);
            
            // Remove text in asterisks (e.g., "*cough*")
            result = RemoveAsterisks(result);
            
            // Normalize capitalization (e.g., "HeLLo" -> "Hello")
            result = NormalizeCasing(result);
            
            return result;
        }

        /// <summary>
        /// Normalizes repeated characters in text.
        /// </summary>
        /// <param name="input">The text to normalize.</param>
        /// <returns>The text with normalized repeated characters.</returns>
        public string NormalizeRepeatedCharacters(string input)
        {
            if (string.IsNullOrEmpty(input))
                return input;

            // Define word boundaries to identify words
            string pattern = @"\b\w+\b";
            
            return Regex.Replace(input, pattern, match =>
            {
                string word = match.Value;
                StringBuilder normalized = new StringBuilder();
                
                for (int i = 0; i < word.Length; i++)
                {
                    char current = word[i];
                    normalized.Append(current);
                    
                    // Skip all subsequent identical characters
                    while (i + 1 < word.Length && word[i + 1] == current)
                    {
                        i++;
                    }
                }
                
                return normalized.ToString();
            });
        }

        /// <summary>
        /// Removes text surrounded by asterisks and fixes any resulting extra spaces.
        /// </summary>
        /// <param name="input">The text to normalize.</param>
        /// <returns>The text without asterisk content and no extra spaces.</returns>
        public string RemoveAsterisks(string input)
        {
            if (string.IsNullOrEmpty(input))
                return input;
            
            // Remove text between asterisks completely
            string result = Regex.Replace(input, @"\*[^*]*\*", "");
            
            // Fix any resulting double spaces or more
            result = Regex.Replace(result, @"\s{2,}", " ");
            
            return result;
        }

        /// <summary>
        /// Normalizes capitalization of words.
        /// </summary>
        /// <param name="input">The text to normalize.</param>
        /// <returns>The text with normalized capitalization.</returns>
        public string NormalizeCasing(string input)
        {
            if (string.IsNullOrEmpty(input))
                return input;

            // Define word boundaries to identify words
            string pattern = @"\b\w+\b";
            
            return Regex.Replace(input, pattern, match =>
            {
                string word = match.Value;
                
                // If the word has mixed capitalization
                if (HasMixedCasing(word))
                {
                    // Convert everything to lowercase
                    string normalized = word.ToLower();
                    
                    // First letter uppercase if the original was uppercase
                    if (char.IsUpper(word[0]))
                    {
                        normalized = char.ToUpper(normalized[0]) + normalized.Substring(1);
                    }
                    
                    return normalized;
                }
                
                return word;
            });
        }

        /// <summary>
        /// Checks if a word contains mixed capitalization.
        /// </summary>
        /// <param name="word">The word to check.</param>
        /// <returns>True if the word has mixed capitalization, otherwise False.</returns>
        private bool HasMixedCasing(string word)
        {
            if (word.Length <= 1)
                return false;
            
            bool hasUpper = false;
            bool hasLower = false;
            
            for (int i = 0; i < word.Length; i++)
            {
                if (char.IsUpper(word[i]))
                    hasUpper = true;
                else if (char.IsLower(word[i]))
                    hasLower = true;
                
                if (hasUpper && hasLower)
                    return true;
            }
            
            return false;
        }
    }
}